File: sdk/src/main/java/com/robotemi/sdk/BatteryData.java
package com.robotemi.sdk;

import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.NonNull;

public class BatteryData implements Parcelable {

    @NonNull
    private final int level;

    @NonNull
    private final boolean isCharging;

    public BatteryData(@NonNull final int level, @NonNull final boolean isCharging) {
        this.level = level;
        this.isCharging = isCharging;
    }

    protected BatteryData(Parcel in) {
        level = in.readInt();
        isCharging = in.readInt() == 1;
    }

    public static final Creator<BatteryData> CREATOR = new Creator<BatteryData>() {
        @Override
        public BatteryData createFromParcel(Parcel in) {
            return new BatteryData(in);
        }

        @Override
        public BatteryData[] newArray(int size) {
            return new BatteryData[size];
        }
    };

    @NonNull
    public int getBatteryPercentage() {
        return level;
    }

    @NonNull
    public boolean isCharging() {
        return isCharging;
    }

    @Override
    public String toString() {
        return "BatteryData{" +
                "level=" + level +
                ", isCharging=" + isCharging +
                '}';
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(level);
        dest.writeInt(isCharging ? 1 : 0);
    }
}

File: sdk/src/main/java/com/robotemi/sdk/BatteryData.kt
package com.robotemi.sdk

import android.os.Parcel
import android.os.Parcelable

/**
 * @property battery2Level Supported in 134 version.
 *                         On some temi built from platform, there will be an extra battery dedicated to power the add-on hardware.
 *                         This property will be used to indicate the battery level of it.
 */
data class BatteryData(
    @get: JvmName("getBatteryPercentage")
    val level: Int,
    val isCharging: Boolean,
    private val hasBattery2: Boolean,
    private val _battery2Level: Int,
) : Parcelable {

    val battery2Level: Int?
        get() = if (hasBattery2) _battery2Level else null

    constructor(source: Parcel) : this(
        source.readInt(),
        source.readInt() == 1,
        source.readInt() == 1,
        source.readInt(),
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeInt(level)
        writeInt(if (isCharging) 1 else 0)
        writeInt(if (hasBattery2) 1 else 0)
        writeInt(_battery2Level)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<BatteryData> = object : Parcelable.Creator<BatteryData> {
            override fun createFromParcel(source: Parcel): BatteryData = BatteryData(source)
            override fun newArray(size: Int): Array<BatteryData?> = arrayOfNulls(size)
        }
    }
}